featurebyte/models/entity_universe.py

```diff
@@ -0,0 +1,334 @@
+    aggregate_result_table_names: List[str]
+        The names of the aggregate result tables
+
+    Returns
+    -------
+    Expression
+    """
+    if not serving_names:
+        return expressions.select(
+            expressions.alias_(make_literal_value(1), "dummy_entity", quoted=True)
+        )
+
+    assert len(aggregate_result_table_names) > 0
+
+    # select distinct serving names across all aggregation result tables
+    distinct_serving_names_from_tables = [
+        select(*[quoted_identifier(serving_name) for serving_name in serving_names])
+        .distinct()
+        .from_(table_name)
+        for table_name in aggregate_result_table_names
+    ]
+
+    union_expr = distinct_serving_names_from_tables[0]
+    for expr in distinct_serving_names_from_tables[1:]:
+        union_expr = expressions.Union(this=expr, distinct=True, expression=union_expr)  # type: ignore
+
+    return union_expr
+
+
+class EntityUniverseModel(FeatureByteBaseModel):
+    """
+    EntityUniverseModel class
+    """
+
+    # query_template is a SQL expression template with placeholders __fb_current_feature_timestamp,
+    # __fb_last_materialized_timestamp which will be replaced with actual values when entity
+    # universe needs to be generated.
+    query_template: SqlglotExpressionModel
+
+    @classmethod
+    def create(
+        cls,
+        serving_names: List[str],
+        aggregate_result_table_names: List[str],
+        offline_ingest_graphs: List[OfflineStoreIngestQueryGraph],
+        source_type: SourceType,
+    ) -> EntityUniverseModel:
+        """
+        Create a new EntityUniverseModel object
+
+        Parameters
+        ----------
+        serving_names: List[str]
+            The serving names of the entities
+        aggregate_result_table_names: List[str]
+            The names of the aggregate result tables for window aggregates
+        offline_ingest_graphs: List[OfflineStoreIngestQueryGraph]
+            The offline ingest graphs that the entity universe is to be constructed from
+        source_type: SourceType
+            Source type information
+
+        Returns
+        -------
+        EntityUniverse
+        """
+        if aggregate_result_table_names:
+            universe_expr = construct_window_aggregates_universe(
+                serving_names=serving_names,
+                aggregate_result_table_names=aggregate_result_table_names,
+            )
+        else:
+            graph_and_node_pairs = []
+            for offline_ingest_graph in offline_ingest_graphs:
+                for info in offline_ingest_graph.aggregation_nodes_info:
+                    node = offline_ingest_graph.graph.get_node_by_name(info.node_name)
+                    graph_and_node_pairs.append((offline_ingest_graph.graph, node))
+            universe_expr = get_combined_universe(graph_and_node_pairs, source_type)
+
+        return EntityUniverseModel(query_template=SqlglotExpressionModel.create(universe_expr))
```

Comment:
If i understood it correctly, we only expect one of `aggregate_result_table_names` or `offline_ingest_graphs` is provided when calling this construction method. Would be good to add assertions to check this.

Reply:
Yes, that's right. Will add assertions in a follow up.